File: api/internal/ivf/kernel/top5.go
// Package kernel implements the inner-loop distance scan for IVF search.
// Two implementations are selected at build time:
//
//   - cgo + amd64:    AVX2 SIMD via kernel_amd64.c (fast)
//   - everything else: pure Go scalar fallback
//
// Both update a shared Top5 candidate list in place.
package kernel

// K and D mirror the constants in the parent IVF package and intentionally
// duplicate them so this package can stand on its own.
const (
	K = 5  // top-K nearest neighbors
	D = 14 // vector dimension
)

// Top5 holds the running top-K candidates. Layout is fixed and matches the
// equivalent C struct sketchy_top5_t in kernel_amd64.h byte-for-byte so the
// cgo wrapper can hand the same memory to C without copying.
//
// Total size: 88 bytes, alignment 8.
//
// Tie-break rule (matches the C reference): a candidate (d, id) beats (d', id')
// iff d < d' OR (d == d' AND id < id'). This makes the top-5 deterministic
// against the brute-force grader.
type Top5 struct {
	BestD   [5]uint64 // squared distances, each <= ~14 * (20000)^2 = 5.6e9, fits uint64 trivially
	BestID  [5]uint32 // orig_id (record's original position in references.json.gz)
	BestL   [5]uint8  // 1 = fraud, 0 = legit
	_pad    [3]uint8
	Worst   int32  // index in BestD/BestID/BestL of the current worst candidate
	WorstD  uint64 // BestD[Worst], cached for fast comparison in hot loop
	WorstID uint32 // BestID[Worst], cached
	_pad2   uint32
}

// Reset initializes top5 to an empty state ready for a fresh query.
// All slots start as "infinitely bad" so the first 5 valid candidates
// will replace them in any order, then the loop converges to true top-5.
func (t *Top5) Reset() {
	for i := 0; i < 5; i++ {
		t.BestD[i] = ^uint64(0)
		t.BestID[i] = ^uint32(0)
		t.BestL[i] = 0
	}
	t.Worst = 0
	t.WorstD = ^uint64(0)
	t.WorstID = ^uint32(0)
}

// Frauds returns the count of fraud labels among the top-5.
func (t *Top5) Frauds() int {
	n := 0
	for i := 0; i < 5; i++ {
		if t.BestL[i] == 1 {
			n++
		}
	}
	return n
}

File: api-rust/src/kernel/top5.rs
// Top5 — the running k=5 nearest-neighbor list. Layout and tie-break rule
// match api/internal/ivf/kernel/top5.go and kernel_amd64.c byte-for-byte.

#[derive(Clone, Copy)]
pub struct Top5 {
    pub best_d: [u64; 5],
    pub best_id: [u32; 5],
    pub best_l: [u8; 5],
    pub worst: usize,
    pub worst_d: u64,
    pub worst_id: u32,
}

impl Top5 {
    pub fn new() -> Self {
        Self {
            best_d: [u64::MAX; 5],
            best_id: [u32::MAX; 5],
            best_l: [0; 5],
            worst: 0,
            worst_d: u64::MAX,
            worst_id: u32::MAX,
        }
    }

    #[inline]
    pub fn frauds(&self) -> i32 {
        let mut n = 0;
        for i in 0..5 {
            if self.best_l[i] == 1 {
                n += 1;
            }
        }
        n
    }

    /// Insert `(d, label, oid)` if it beats the current worst slot. Tie-break:
    /// `(d, id)` lex order (smaller wins), matching the C/Go reference.
    #[inline]
    pub fn try_insert(&mut self, d: u64, label: u8, oid: u32) {
        if !is_better(d, oid, self.worst_d, self.worst_id) {
            return;
        }
        let w = self.worst;
        self.best_d[w] = d;
        self.best_l[w] = label;
        self.best_id[w] = oid;
        let nw = find_worst(&self.best_d, &self.best_id);
        self.worst = nw;
        self.worst_d = self.best_d[nw];
        self.worst_id = self.best_id[nw];
    }
}

#[inline]
pub fn is_better(da: u64, ia: u32, db: u64, ib: u32) -> bool {
    da < db || (da == db && ia < ib)
}

#[inline]
pub fn is_worse(da: u64, ia: u32, db: u64, ib: u32) -> bool {
    da > db || (da == db && ia > ib)
}

#[inline]
fn find_worst(d: &[u64; 5], id: &[u32; 5]) -> usize {
    let mut w = 0;
    if is_worse(d[1], id[1], d[w], id[w]) {
        w = 1;
    }
    if is_worse(d[2], id[2], d[w], id[w]) {
        w = 2;
    }
    if is_worse(d[3], id[3], d[w], id[w]) {
        w = 3;
    }
    if is_worse(d[4], id[4], d[w], id[w]) {
        w = 4;
    }
    w
}
